src/lib/AllocationCbor.sol

```diff
@@ -0,0 +1,184 @@
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.25;
+
+import "@ensdomains/buffer/contracts/Buffer.sol"; // for appendUint8
+import {CBOR} from "solidity-cborutils/contracts/CBOR.sol";
+import {CBORDecoder} from "filecoin-solidity/utils/CborDecode.sol";
+
+struct AllocationRequestData {
+    uint64 provider;
+    bytes dataCID;
+    uint64 size;
+    int64 termMin;
+    int64 termMax;
+    int64 expiration;
+}
+
+library AllocationCbor {
+    using Buffer for Buffer.buffer; // for appendUint8
+
+    function encodeData(
+        AllocationRequestData memory request
+    ) internal pure returns (bytes memory) {
+        CBOR.CBORBuffer memory buf = CBOR.create(64);
+
+        // Start the top-level array
+        CBOR.startFixedArray(buf, 2);
+
+        // Start the first nested array
+        CBOR.startFixedArray(buf, 1);
+
+        // Start the allocation request array
+        CBOR.startFixedArray(buf, 6);
+        CBOR.writeUInt64(buf, request.provider);
+        // Insert tag(42) before writing the dataCID
+        writeTag(buf, 42);
+        CBOR.writeBytes(buf, request.dataCID);
+        CBOR.writeUInt64(buf, request.size);
+        CBOR.writeInt64(buf, request.termMin);
+        CBOR.writeInt64(buf, request.termMax);
+        CBOR.writeInt64(buf, request.expiration);
+
+        // Write the empty array
+        CBOR.startFixedArray(buf, 0);
+
+        return CBOR.data(buf);
+    }
+
+    function encodeDataArray(
+        AllocationRequestData[] memory requests
+    ) internal pure returns (bytes memory) {
+        // Create a new CBOR buffer with an initial capacity.
+        CBOR.CBORBuffer memory buf = CBOR.create(128);
```

Comment:
This should probably be smarter. Something like `requests.length * 64 + 3`? Not sure about the `3`, as I think it might encode the array length in more bytes if it exceeds some length.

Reply:
And we should make some effort to get this right, as reallocations will be expensive.

Reply:
Seem like this can help: https://github.com/filecoin-project/filecoin-solidity/blob/master/contracts/v0.8/utils/Misc.sol#L61